flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/DataSetConversionUtil.java

```diff
@@ -0,0 +1,171 @@
+			.get(sessionId)
+			.getBatchTableEnvironment()
+			.toDataSet(table, Row.class);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified TableSchema.
+	 *
+	 * @param sessionId the sessionId of {@link MLEnvironmentFactory}
+	 * @param data   the DataSet to convert.
+	 * @param schema the specified TableSchema.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, TableSchema schema) {
+		return toTable(sessionId, data, schema.getFieldNames(), schema.getFieldTypes());
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames and colTypes.
+	 *
+	 * @param sessionId sessionId the sessionId of {@link MLEnvironmentFactory}.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @param colTypes the specified colTypes. This variable is used only when the
+	 *                 DataSet is produced by a function and Flink cannot determine
+	 *                 automatically what the produced type is.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, String[] colNames, TypeInformation <?>[] colTypes) {
+		return toTable(MLEnvironmentFactory.get(sessionId), data, colNames, colTypes);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames.
+	 *
+	 * @param sessionId sessionId the sessionId of {@link MLEnvironmentFactory}.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, String[] colNames) {
+		return toTable(MLEnvironmentFactory.get(sessionId), data, colNames);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames and colTypes.
+	 *
+	 * @param session the MLEnvironment using to convert DataSet to Table.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @param colTypes the specified colTypes. This variable is used only when the
+	 *                 DataSet is produced by a function and Flink cannot determine
+	 *                 automatically what the produced type is.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(MLEnvironment session, DataSet <Row> data, String[] colNames, TypeInformation <?>[] colTypes) {
+		try {
+			// Try to add row type information for the dataset to be converted.
+			// In most case, this keeps us from the rolling back logic in the catch block,
+			// which adds an unnecessary map function just in order to add row type information.
+			if (data instanceof SingleInputUdfOperator) {
+				((SingleInputUdfOperator) data).returns(new RowTypeInfo(colTypes, colNames));
+			} else if (data instanceof TwoInputUdfOperator) {
+				((TwoInputUdfOperator) data).returns(new RowTypeInfo(colTypes, colNames));
+			}
+			return toTable(session, data, colNames);
+		} catch (Exception ex) {
```

Comment:
Is there anyway to avoid doing this. Ideally speaking an exception is an exception and we shouldn't catch it and branch out to a different solution.

I am assuming you can do some pre-check to see if this `toTable` operation is valid (If I understood correctly, basically doing the type check between the `colTypes` and the actual dataset type right?)

Reply:
also this exception branch is not tested.

Reply:
because getType() in https://github.com/apache/flink/blob/937356f786b5edac28a7ff44fc31410761c6d7f0/flink-java/src/main/java/org/apache/flink/api/java/DataSet.java#L171 can only be called once. so we can not do the pre-check.

Reply:
I am actually a little confused on this. why is the `getType()` not able to be used more than once? It seems like you cannot alter the `typeInfo` once you have access the `getType()` API. 

however you can still access the type again yes? Also in the exception handle mechanism, `getDataSetWithExplicitTypeDefine` method, you are actually making a `map` first before you assign the return type?

Reply:
Also, it is probably better to list out the actual Exception you are trying to catch here instead of a generic `Exception` and then handle it. It is unknown to contributors who maintain the `DataSet` side code --> if they change and throw some new Exceptions that happens to cross path for this method, it will silently fail (since there's not test that actually touch the `catch` branch.

Reply:
The actual exception thrown here is this `ValidationException`:
https://github.com/apache/flink/blob/ef4055ab989db431e29e0071573619d267960000/flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/typeutils/FieldInfoUtils.java#L184-L189
That's because sometimes the type information of Row bassed dataset is not set.

As you mentioned, our initial solution is doing some pre-check, saying checking if the `TypeInformation` is `GenericTypeInfo<Row>`. To check the type, we need to `getType()`. As a side effect, `typeUsed` is set to `true`:

https://github.com/apache/flink/blob/937356f786b5edac28a7ff44fc31410761c6d7f0/flink-java/src/main/java/org/apache/flink/api/java/DataSet.java#L171-L182

 But what if the input dataset really has no specified type information? We have to set it's TypeInformation via the `.return()` methods, which calls the following code, which says "no modification after getType called":
https://github.com/apache/flink/blob/937356f786b5edac28a7ff44fc31410761c6d7f0/flink-java/src/main/java/org/apache/flink/api/java/DataSet.java#L155-L161

That's the deadlock we faced. So we just try the best to complete the type information (for `SingleInputUdfOperator` and `TwoInputUdfOperator` here) to avoid falling into the exception branch and let the `getDataSetWithExplicitTypeDefine` method work as a final guard. 

BTW, catching `ValidationException` instead of a generic `Exception` is a good idea I think.

Reply:
This seems to me is like a bug in `DataSet` - if an exception was thrown, then `typeUsed` should not be set (since an exception was actually thrown). 

However, I think I understand the situation a lot better, thanks for the clarification @qiuxiafei. I would dig deeper in the DataSet/TypeInfo side code and see if there's any better solution.